scripts/serialization/compare.py

```diff
@@ -0,0 +1,283 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Compare dataset serialization methods.
+
+Serialization methods compared:
+* Arrow https://arrow.apache.org/
+* MDS https://github.com/mosaicml/streaming
+* Parquet https://parquet.apache.org/
+
+We generate datasets containing identical samples in each format, and compare the time it takes to
+iterate over them in (a) sequential order and (b) shuffled order.
+
+We find that MDS has the best random access performance, identical to its sequential access. This
+is critical because machine learning training jobs (its intended use case) iterate over the
+training split in shuffled order.
+"""
+
+import json
+import os
+from argparse import ArgumentParser, Namespace
+from glob import glob
+from shutil import rmtree
+from time import time
+from typing import Any, Callable, Dict, Iterator, List
+
+import numpy as np
+import pandas as pd
+from datasets import Dataset, disable_caching, load_dataset, load_from_disk  # pyright: ignore
+from matplotlib import pyplot as plt
+from tqdm import tqdm
+
+from streaming import MDSWriter, StreamingDataset
+
+disable_caching()
+
+
+def parse_args() -> Namespace:
+    """Parse command-line arguments.
+
+    Returns:
+        Namespace: Command-line arguments.
+    """
+    args = ArgumentParser()
+    args.add_argument('--data', type=str, default='/tmp/streaming/perf_test/')
+    args.add_argument('--num_samples', type=int, default=200_000)
+    args.add_argument('--out', type=str, default='plot.png')
+    return args.parse_args()
+
+
+def fetch_dataset(data_root: str, num_samples: int) -> List[Dict[str, Any]]:
+    """Fetch dataset to benchmark as JSONL.
+
+    Args:
+        data_root (str): Directory to cache samples in.
+        num_samples (int): Only use the first this many samples.
+
+    Returns:
+        List[Dict[str, Any]]: The samples.
+    """
+    if not os.path.isdir(data_root):
+        os.makedirs(data_root)
+    filename = os.path.join(data_root, 'samples.jsonl')
+
+    samples = []
+    if os.path.exists(filename):
+        for i, line in tqdm(enumerate(open(filename)), total=num_samples, leave=False):
+            if i == num_samples:
+                break
+            sample = json.loads(line)
+            samples.append(sample)
+
+        if len(samples) != num_samples:
+            samples = []
+            os.remove(filename)
+
+    if not os.path.exists(filename):
+        dataset = load_dataset('the_pile', 'pubmed_central', split='train', streaming=True)
+        with open(filename, 'w') as out:
+            for i, sample in tqdm(
+                    enumerate(dataset),
+                    total=len(dataset),  # pyright: ignore
```

Comment:
I get a `TypeError: object of type 'IterableDataset' has no len()` here when running locally (I don't think len will work on a HF streaming dataset) ... could probably use `total=num_samples`